src/transaction/atomic_replication_helper.cpp

```diff
@@ -363,33 +365,26 @@ namespace cublog
     return false;
   }
 
-  void
-  atomic_replication_helper::atomic_log_sequence::apply_all_log_redos (THREAD_ENTRY *thread_p)
-  {
-    for (size_t i = 0; i < m_log_vec.size (); i++)
-      {
-	m_log_vec[i].apply_log_redo (thread_p, m_redo_context);
-      }
-  }
-
   void
   atomic_replication_helper::atomic_log_sequence::apply_and_unfix_sequence (THREAD_ENTRY *thread_p)
   {
     // Applying the log right after the fix could lead to problems as the records are fixed one by one as
     // they come to be read by the PTS and some might be unfixed and refixed after the apply procedure
-    // leading to inconsistency. To avoid this situation we sequentially apply each log redo of the sequence
-    // when the end sequence log appears and the entire sequence is fixed
-    apply_all_log_redos (thread_p);
-
-    for (size_t i = 0; i < m_log_vec.size (); i++)
+    // leading to inconsistency.
+    // To avoid this situation each log redo of the sequence is applied when the end sequence log appears
+    // and the entire sequence is already fixed.
+    // Right after applying, unfix and ref-count-down each page. The bookkeeping mechanism will take care
+    // of either unfixing the page or retaining it for a subsequent unfix.
+    for (const auto &log_entry : m_log_vec)
       {
-	const auto page_map_it = m_page_map.find (m_log_vec[i].m_vpid);
-	if (page_map_it != m_page_map.end ())
-	  {
-	    m_log_vec[i].unfix_page (thread_p);
-	    m_page_map.erase (page_map_it);
-	  }
+	log_entry.apply_log_redo (thread_p, m_redo_context);
+	// bookkeeping actually will either unfix the page or just decrease its reference count
+	m_page_ptr_bookkeeping.unfix_page (thread_p, log_entry.m_vpid);
```

Comment:
How about doing unifx on atomic_log_entry's destructor and fix on its constructor in terms of RAII? It seems that fixing and unfixing always go along with the atomic_log_entry's life cycle.

Reply:
On the fix/new entry side, they are decoupled in the sense that an error can interrupt the addition of a new entry - but that is an error case.
On the unfix/dtor side they are one and the same indeed.
I will leave it as it is. The sky is the limit in terms of C++ abstraction anyway and there is no clear line where one should stop or not. :)

Reply:
Ok. Thank you for considering it.